libpam/pam_handlers.c

```diff
@@ -572,93 +584,166 @@ int _pam_init_handlers(pam_handle_t *pamh)
-    char *s, *os;
     int used = 0;
 
     /* loop broken with a 'break' when a non-'\\n' ended line is read */
 
     D(("called."));
+
+    _pam_buffer_shift(buffer);
+
     for (;;) {
-	if (p >= endp - 1) {
-	    /* Overflow */
-	    D(("overflow"));
-	    return -1;
-	}
-	if (fgets(p, endp - p, f) == NULL) {
+	char *s, *os;
+	ssize_t n;
+
+	if ((n = getline(&buffer->chunk, &buffer->chunk_size, f)) == -1) {
 	    if (used) {
 		/* Incomplete read */
-		return -1;
+		used = -1;
 	    } else {
 		/* EOF */
-		return 0;
+		used = 0;
 	    }
+	    break;
 	}
 
-	if (strchr(p, '\n') == NULL && !feof(f)) {
-	    /* Incomplete */
-	    D(("_pam_assemble_line: incomplete"));
-	    return -1;
-	}
-
-	/* skip leading spaces --- line may be blank */
-
-	s = p + strspn(p, " \n\t");
-	if (*s && (*s != '#')) {
-	    os = s;
-
-	    /*
-	     * we are only interested in characters before the first '#'
-	     * character
-	     */
-
-	    while (*s && *s != '#')
-		 ++s;
-	    if (*s == '#') {
-		 *s = '\0';
-		 used += strlen(os);
-		 break;                /* the line has been read */
-	    }
-
-	    s = os;
-
-	    /*
-	     * Check for backslash by scanning back from the end of
-	     * the entered line, the '\n' has been included since
-	     * normally a line is terminated with this
-	     * character. fgets() should only return one though!
-	     */
-
-	    s += strlen(s);
-	    while (s > os && ((*--s == ' ') || (*s == '\t')
-			      || (*s == '\n')));
-
-	    /* check if it ends with a backslash */
-	    if (*s == '\\') {
-		*s++ = ' ';             /* replace backslash with ' ' */
-		*s = '\0';              /* truncate the line here */
-		used += strlen(os);
-		p = s;                  /* there is more ... */
-	    } else {
-		/* End of the line! */
-		used += strlen(os);
-		break;                  /* this is the complete line */
+	/* skip leading spaces */
+
+	os = s = buffer->chunk + strspn(buffer->chunk, " \t");
+
+	/*
+	 * we are only interested in characters before the first '#'
+	 * character
+	 */
+
+	s = buffer->chunk + strcspn(buffer->chunk, "#");
+	if (*s == '#')
+	    *s = '\0';
+
+	/*
+	 * Check for backslash by scanning back from the end of
+	 * the entered line, the '\n' should be included since
+	 * normally a line is terminated with this character.
+	 */
+
+	s = buffer->chunk + n;
+	while (s > os && ((*--s == ' ') || (*s == '\t') || (*s == '\n')));
+
+	/* check if it ends with a backslash */
+	if (*s == '\\') {
+	    if (buffer->assembled != NULL)
+		*s++ = ' ';         /* replace backslash with ' ' */
+	    *s = '\0';              /* truncate the line here */
+	    if (os != s) {
+		if (_pam_buffer_add(buffer, os, s)) {
+		    used = -1;
+		    break;
+		}
+		used = 1;
+	    }                       /* there is more ... */
+	} else {
+	    /* End of the line! */
+	    if (buffer->assembled != NULL || (*os != '\0' && *os != '\n')) {
+		if (_pam_buffer_add(buffer, os, buffer->chunk + n))
+		    used = -1;
```

Comment:
This `used = -1` has to be followed by `break`, otherwise it's set to 0.

Reply:
Hm, this seems to be a comment for an outdated version. Is this still an issue?

Reply:
The comment is for the penultimate commit in the series.  Apparently, the last commit in the series fixed that, but I'd prefer this to be fixed in the first commit that introduces it.

Reply:
Ah okay, thanks for clarification. Fixed. Please note that we can (and probably) should squash these two commits.

Reply:
Feel free to squash them if you feel they should rather be squashed.

Reply:
I have squashed them now.